### ISSUE TITLE
Typescript throwing errors on latest version and unable to install any previous versions

### ISSUE THREAD

**Issue by realfresh** [ably/ably-js#574]
Hey, not sure if I'm doing something wrong but something weird is happening.

First, since updating to the latest stable version, 1.1.2, when I try to compile my typescript files, I'm getting the following errors below. One extremely strange thing is, I'm completely unable to roll back to any previous versions. If i install any other version through either yarn or npm, it always downloads version 1.1.2 files. Here are the typescript errors

Edit: Seems I was finally able to get Ably to install the right version by removing the package from package.json and then manually installing with "yarn add ably@1.0.23". Starting to feel like this whole problem is probably just something to do with my setup but gonna leave this up just in case.

```
node_modules/ably/browser/static/ably.d.ts:367:3 - error TS2416: Property 'auth' in type 'RestPromise' is not assignable to the same property in base type 'Rest'.
  Type 'AuthPromise' is not assignable to type 'Auth'.
    Types of property 'authorize' are incompatible.
      Type '(tokenParams?: TokenParams | undefined, authOptions?: AuthOptions | undefined) => Promise<TokenDetails>' is not assignable to type '(tokenParams?: TokenParams | tokenDetailsCallback | undefined, authOptions?: AuthOptions | tokenDetailsCallback | undefined, callback?: tokenDetailsCallback | undefined) => void'.
        Types of parameters 'tokenParams' and 'tokenParams' are incompatible.
          Type 'TokenParams | tokenDetailsCallback | undefined' is not assignable to type 'TokenParams | undefined'.
            Type 'tokenDetailsCallback' has no properties in common with type 'TokenParams'.

367     auth: Types.AuthPromise;
        ~~~~

node_modules/ably/browser/static/ably.d.ts:368:3 - error TS2416: Property 'channels' in type 'RestPromise' is not assignable to the same property in base type 'Rest'.
  Type 'Channels<ChannelPromise>' is not assignable to type 'Channels<Channel>'.
    Type 'ChannelPromise' is not assignable to type 'Channel'.
      Types of property 'history' are incompatible.
        Type '(params?: RestHistoryParams | undefined) => Promise<PaginatedResult<Message>>' is not assignable to type '(paramsOrCallback?: RestHistoryParams | paginatedResultCallback<Message> | undefined, callback?: paginatedResultCallback<Message> | undefined) => void'.
          Types of parameters 'params' and 'paramsOrCallback' are incompatible.
            Type 'RestHistoryParams | paginatedResultCallback<Message> | undefined' is not assignable to type 'RestHistoryParams | undefined'.
              Type 'paginatedResultCallback<Message>' has no properties in common with type 'RestHistoryParams'.

368     channels: Types.Channels<Types.ChannelPromise>;
        ~~~~~~~~                                                                                                                          
                                                                                                                                          
node_modules/ably/browser/static/ably.d.ts:376:3 - error TS2416: Property 'auth' in type 'RealtimePromise' is not assignable to the same property in base type 'Realtime'.                                                                                                          
  Type 'AuthPromise' is not assignable to type 'Auth'.                                                                                    
                                                                                                                                          
376     auth: Types.AuthPromise;                                                                                                          
        ~~~~

node_modules/ably/browser/static/ably.d.ts:377:3 - error TS2416: Property 'channels' in type 'RealtimePromise' is not assignable to the same property in base type 'Realtime'.
  Type 'Channels<RealtimeChannelPromise>' is not assignable to type 'Channels<RealtimeChannel>'.
    Type 'RealtimeChannelPromise' is not assignable to type 'RealtimeChannel'.
      Types of property 'history' are incompatible.
        Type '(params?: RealtimeHistoryParams | undefined) => Promise<PaginatedResult<Message>>' is not assignable to type '(paramsOrCallback?: RealtimeHistoryParams | paginatedResultCallback<Message> | undefined, callback?: paginatedResultCallback<Message> | undefined) => void'.
          Types of parameters 'params' and 'paramsOrCallback' are incompatible.
            Type 'RealtimeHistoryParams | paginatedResultCallback<Message> | undefined' is not assignable to type 'RealtimeHistoryParams | undefined'.
              Type 'paginatedResultCallback<Message>' has no properties in common with type 'RealtimeHistoryParams'.

377     channels: Types.Channels<Types.RealtimeChannelPromise>;
        ~~~~~~~~

node_modules/ably/browser/static/ably.d.ts:393:3 - error TS2416: Property 'authorize' in type 'AuthPromise' is not assignable to the same property in base type 'Auth'.
  Type '(tokenParams?: TokenParams | undefined, authOptions?: AuthOptions | undefined) => Promise<TokenDetails>' is not assignable to type '(tokenParams?: TokenParams | tokenDetailsCallback | undefined, authOptions?: AuthOptions | tokenDetailsCallback | undefined, callback?: tokenDetailsCallback | undefined) => void'.

393     authorize: (tokenParams?: TokenParams, authOptions?: AuthOptions) => Promise<TokenDetails>;
        ~~~~~~~~~

node_modules/ably/browser/static/ably.d.ts:394:3 - error TS2416: Property 'createTokenRequest' in type 'AuthPromise' is not assignable to the same property in base type 'Auth'.
  Type '(tokenParams?: TokenParams | undefined, authOptions?: AuthOptions | undefined) => Promise<TokenRequest>' is not assignable to type '(tokenParams?: TokenParams | tokenRequestCallback | undefined, authOptions?: AuthOptions | tokenRequestCallback | undefined, callback?: tokenRequestCallback | undefined) => void'.
    Types of parameters 'tokenParams' and 'tokenParams' are incompatible.
      Type 'TokenParams | tokenRequestCallback | undefined' is not assignable to type 'TokenParams | undefined'.
        Type 'tokenRequestCallback' has no properties in common with type 'TokenParams'.

394     createTokenRequest: (tokenParams?: TokenParams, authOptions?: AuthOptions) => Promise<TokenRequest>;
        ~~~~~~~~~~~~~~~~~~

node_modules/ably/browser/static/ably.d.ts:395:3 - error TS2416: Property 'requestToken' in type 'AuthPromise' is not assignable to the same property in base type 'Auth'.
  Type '(TokenParams?: TokenParams | undefined, authOptions?: AuthOptions | undefined) => Promise<TokenDetails>' is not assignable to type '(TokenParams?: TokenParams | tokenDetailsCallback | undefined, authOptions?: AuthOptions | tokenDetailsCallback | undefined, callback?: tokenDetailsCallback | undefined) => void'.
    Types of parameters 'TokenParams' and 'TokenParams' are incompatible.
      Type 'TokenParams | tokenDetailsCallback | undefined' is not assignable to type 'TokenParams | undefined'.
        Type 'tokenDetailsCallback' has no properties in common with type 'TokenParams'.

395     requestToken: (TokenParams?: TokenParams, authOptions?: AuthOptions) => Promise<TokenDetails>;
        ~~~~~~~~~~~~

node_modules/ably/browser/static/ably.d.ts:404:3 - error TS2416: Property 'get' in type 'PresencePromise' is not assignable to the same property in base type 'Presence'.
  Type '(params?: RestPresenceParams | undefined) => Promise<PaginatedResult<PresenceMessage>>' is not assignable to type '(paramsOrCallback?: RestPresenceParams | paginatedResultCallback<PresenceMessage> | undefined, callback?: paginatedResultCallback<PresenceMessage> | undefined) => void'.
    Types of parameters 'params' and 'paramsOrCallback' are incompatible.
      Type 'RestPresenceParams | paginatedResultCallback<PresenceMessage> | undefined' is not assignable to type 'RestPresenceParams | undefined'.
        Type 'paginatedResultCallback<PresenceMessage>' has no properties in common with type 'RestPresenceParams'.

404     get: (params?: RestPresenceParams) => Promise<PaginatedResult<PresenceMessage>>;
        ~~~

node_modules/ably/browser/static/ably.d.ts:405:3 - error TS2416: Property 'history' in type 'PresencePromise' is not assignable to the same property in base type 'Presence'.
  Type '(params?: RestHistoryParams | undefined) => Promise<PaginatedResult<PresenceMessage>>' is not assignable to type '(paramsOrCallback: RestHistoryParams | paginatedResultCallback<PresenceMessage>, callback?: paginatedResultCallback<PresenceMessage> | undefined) => void'.
    Types of parameters 'params' and 'paramsOrCallback' are incompatible.
      Type 'RestHistoryParams | paginatedResultCallback<PresenceMessage>' is not assignable to type 'RestHistoryParams | undefined'.
        Type 'paginatedResultCallback<PresenceMessage>' has no properties in common with type 'RestHistoryParams'.

405     history: (params?: RestHistoryParams) => Promise<PaginatedResult<PresenceMessage>>;
        ~~~~~~~

node_modules/ably/browser/static/ably.d.ts:423:3 - error TS2416: Property 'get' in type 'RealtimePresencePromise' is not assignable to the same property in base type 'RealtimePresence'.
  Type '(params?: RealtimePresenceParams | undefined) => Promise<PaginatedResult<PresenceMessage>>' is not assignable to type '(paramsOrCallback?: RealtimePresenceParams | realtimePresenceGetCallback | undefined, callback?: realtimePresenceGetCallback | undefined) => void'.
    Types of parameters 'params' and 'paramsOrCallback' are incompatible.
      Type 'RealtimePresenceParams | realtimePresenceGetCallback | undefined' is not assignable to type 'RealtimePresenceParams | undefined'.
        Type 'realtimePresenceGetCallback' has no properties in common with type 'RealtimePresenceParams'.

423     get: (params?: RealtimePresenceParams) => Promise<PaginatedResult<PresenceMessage>>;
        ~~~

node_modules/ably/browser/static/ably.d.ts:424:3 - error TS2416: Property 'history' in type 'RealtimePresencePromise' is not assignable to the same property in base type 'RealtimePresence'.
  Type '(params?: RealtimeHistoryParams | undefined) => Promise<PaginatedResult<PresenceMessage>>' is not assignable to type '(paramsOrCallback?: RealtimeHistoryParams | paginatedResultCallback<PresenceMessage> | undefined, callback?: paginatedResultCallback<PresenceMessage> | undefined) => void'.
    Types of parameters 'params' and 'paramsOrCallback' are incompatible.
      Type 'RealtimeHistoryParams | paginatedResultCallback<PresenceMessage> | undefined' is not assignable to type 'RealtimeHistoryParams | undefined'.
        Type 'paginatedResultCallback<PresenceMessage>' has no properties in common with type 'RealtimeHistoryParams'.

424     history: (params?: RealtimeHistoryParams) => Promise<PaginatedResult<PresenceMessage>>;
        ~~~~~~~

node_modules/ably/browser/static/ably.d.ts:442:3 - error TS2416: Property 'history' in type 'ChannelPromise' is not assignable to the same property in base type 'Channel'.
  Type '(params?: RestHistoryParams | undefined) => Promise<PaginatedResult<Message>>' is not assignable to type '(paramsOrCallback?: RestHistoryParams | paginatedResultCallback<Message> | undefined, callback?: paginatedResultCallback<Message> | undefined) => void'.

442     history: (params?: RestHistoryParams) => Promise<PaginatedResult<Message>>;
        ~~~~~~~

node_modules/ably/browser/static/ably.d.ts:464:3 - error TS2416: Property 'history' in type 'RealtimeChannelPromise' is not assignable to the same property in base type 'RealtimeChannel'.
  Type '(params?: RealtimeHistoryParams | undefined) => Promise<PaginatedResult<Message>>' is not assignable to type '(paramsOrCallback?: RealtimeHistoryParams | paginatedResultCallback<Message> | undefined, callback?: paginatedResultCallback<Message> | undefined) => void'.

464     history: (params?: RealtimeHistoryParams) => Promise<PaginatedResult<Message>>;
        ~~~~~~~


Found 13 errors.
```

**Comment by SimonWoolf** [comment]
Hi @realfresh. So it looks like typescript objects to the way I added Promise support to the ably-js 1.1 typings, when run with `--strictFunctionTypes` (or `--strict`, which implies `--strictFunctionTypes`). I missed that as I was running tsc without that flag -- my bad.

(Specifically, it doesn't like that the promisified versions of the classes, which inherit from the standard ones, change the return type, to return a promise)

I will rewrite the typings so they validate with `--strict`. Thanks for reporting :slightly_smiling_face:

**Comment by SimonWoolf** [comment]
@realfresh this should be fixed in v1.1.3 of the library. Thanks again :slightly_smiling_face:
